fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java

```diff
@@ -153,6 +172,81 @@ private void logMVParams(ConnectContext connectContext, Set<Table> queryTables)
                 sessionVariable.getMaterializedViewRewriteMode());
         logMVPrepare(connectContext, "---------------------------------");
     }
+
+    private void processPlanWithView(ConnectContext connectContext,
+                                     OptExpression logicOperatorTree,
+                                     ColumnRefFactory columnRefFactory,
+                                     ColumnRefSet requiredColumns) {
+        Map<LogicalViewScanOperator, OptExpression> viewPlanMap = Maps.newHashMap();
+        // process equivalent operator，construct logical plan with view
+        OptExpression logicalPlanWithView = extractLogicalPlanWithView(logicOperatorTree, viewPlanMap);
+        if (viewPlanMap.isEmpty()) {
+            // means there is no plan with view
+            return;
+        }
+        // optimize logical plan with view
+        OptExpression optimizedPlan = optimizeViewPlan(
+                logicalPlanWithView, connectContext, requiredColumns, columnRefFactory);
+        context.setLogicalTreeWithView(optimizedPlan);
+
+        Map<LogicalViewScanOperator, OptExpression> optimizedViewPlanMap = Maps.newHashMap();
+        for (LogicalViewScanOperator viewScanOperator : viewPlanMap.keySet()) {
+            OptExpression viewLogicalTree = viewPlanMap.get(viewScanOperator);
+            // optimize logical tree of view and keep them in OptimizerContext,
+            // which will be used in union rewrite or only some views are rewritten
+            OptExpression optimizedViewPlan = optimizeViewPlan(
+                    viewLogicalTree, connectContext, viewScanOperator.getOutputColumnSet(), columnRefFactory);
+            optimizedViewPlanMap.put(viewScanOperator, optimizedViewPlan);
+        }
+        context.setViewPlanMap(optimizedViewPlanMap);
+    }
+
+    private OptExpression optimizeViewPlan(
+            OptExpression logicalTree,
+            ConnectContext connectContext,
+            ColumnRefSet requiredColumns,
+            ColumnRefFactory columnRefFactory) {
+        OptimizerConfig optimizerConfig = new OptimizerConfig(OptimizerConfig.OptimizerAlgorithm.RULE_BASED);
+        optimizerConfig.disableRuleSet(RuleSetType.SINGLE_TABLE_MV_REWRITE);
+        optimizerConfig.disableRuleSet(RuleSetType.MULTI_TABLE_MV_REWRITE);
+        Optimizer optimizer = new Optimizer(optimizerConfig);
+        OptExpression optimizedViewPlan = optimizer.optimize(connectContext, logicalTree,
+                new PhysicalPropertySet(), requiredColumns, columnRefFactory);
+        return optimizedViewPlan;
+    }
+
+    private OptExpression extractLogicalPlanWithView(
```

Comment:
one view can nests another,  this functions only take into consideration the highest view in the plan? is it reasonable?

Reply:
now we only take the most top views in the plan. It can be optimized later, it can be extend without too many modification. But it will incur more search space and more costs. and it should use memo, and it will be more unstable.